### PR TITLE
chore(flushMicroTasks): Use the mocked scheduler when available

### PR DESCRIPTION
**Pull request by MatanBobi**
Following a conversation with @eps1lon, he pointed out that maybe it would be better to mock the Scheduler (based on [this](https://github.com/facebook/react/pull/14964) PR by @acdlite).

This solution will handle both mocked version of scheduler and original version.

**What**: I've mocked the scheduler package in our setup-env and used the `flushAll` function which is available in the mocked scheduler package.

**Why**: React core team recommends to use a mocked version of the scheduler in test environments. We need to think if we want to recommend our users to mock the scheduler package also in their test environments. I think that if our users will mock the scheduler package, we won't need to know if someone is using fake timers or not and just call `flushAll`.

**How**: Checked for the presence of `unstable_flushAll` and used it if it exists. 

**Checklist**:

- [ ] Documentation added to the
      [docs site](https://github.com/testing-library/testing-library-docs) - N/A
- [ ] Tests - N/A
- [ ] Typescript definitions updated - N/A
- [ ] Ready to be merged
      

I'm opening this one to create the conversation here and see what you think about it.

### CODE REVIEW

File: src/flush-microtasks.js [hunk 1]
```diff
@@ -86,6 +87,9 @@ export default function flushMicroTasks() {
         // reproduce the problem, so there's no test for it. But it works!
         jest.advanceTimersByTime(0)
         resolve()
+      } else if (Scheduler && Scheduler.unstable_flushAll) {
+        Scheduler.unstable_flushAll()
+        enqueueTask(resolve)
       } else {
         scheduleCallback(() => enqueueTask(resolve))
       }
```
Comment (kentcdodds) [inline src/flush-microtasks.js:95]:
```suggestion
      } else {
        Scheduler?.unstable_flushAll?.()
        enqueueTask(resolve)
      }
```

Comment (MatanBobi) [reply src/flush-microtasks.js:95]: I think that in that case we won't `scheduleCallback` in case we're not using the mocked scheduler, right?

**Comment by MatanBobi** [comment]
Thanks @kentcdodds! 
Do you have any thoughts on whether we should recommend our users to mock the scheduler package or not?

File: src/flush-microtasks.js [hunk 2]
```diff
@@ -86,6 +87,9 @@ export default function flushMicroTasks() {
         // reproduce the problem, so there's no test for it. But it works!
         jest.advanceTimersByTime(0)
         resolve()
+      } else if (Scheduler && Scheduler.unstable_flushAll) {
+        Scheduler.unstable_flushAll()
+        enqueueTask(resolve)
```
Comment (eps1lon) [inline src/flush-microtasks.js:92]: We don't have to queue the task. Should be able resolve immediately.

Comment (MatanBobi) [reply src/flush-microtasks.js:92]: I tried that, I wasn't able to understand why it didn't work and didn't get the time to look into it yesterday,
I'll look into it today.

Comment (eps1lon) [reply src/flush-microtasks.js:92]: Are we supposed to wait for microtasks in cleanup functions? Does it even make sense to run async code in the cleanup function?

Comment (kentcdodds) [reply src/flush-microtasks.js:92]: I'd we can avoid this being async that would be awesome, but I don't think we can avoid it because we can't assume people will be mocking the scheduler (especially when they're testing in environments with poor support for mocking).

I assumed that calling flushAll would mean we wouldn't need to worry about scheduling a callback. Sure wish we had a test that reproduces these timing issues. I have one in my bookshelf app that have me trouble before making all these changes. I can try these changes out there. I just couldn't simplify it to a reproducible example for inclusion in this project 😬

**Comment by kentcdodds** [comment]
I think this is the right approach: https://github.com/testing-library/react-testing-library/pull/744

File: src/flush-microtasks.js (hunk 2 above)
Comment (eps1lon) [reply src/flush-microtasks.js:92]: One suspicion I have is that you clean up your timers after the cleanup from testing-library runs. At least this is one issue I encountered in my mocha tests. Basically if you import from testing-libary we trigger an afterEach hook with cleanup and then in your test you call afterEach to cleanup timers. But this results in the cleanup of testing-libary being called with fake timers enabled which is fundamentally flawed (people can't expect to check for fake timers since they're not standardized).

If you can point me to the test in question I could check out if that is indeed the case.

**Comment by MatanBobi** [comment]
> I think this is the right approach: #744

Yes, definitely. I'm closing this one. Thanks!

File: src/flush-microtasks.js (hunk 2 above)
Comment (kentcdodds) [reply src/flush-microtasks.js:92]: The test that caused the problem was this one: https://github.com/kentcdodds/bookshelf/blob/85e2880abb3d3290c45a5a0ce061d3e8197b395a/src/__tests__/book-screen.js#L154-L175

But the test that failed was this one: https://github.com/kentcdodds/bookshelf/blob/85e2880abb3d3290c45a5a0ce061d3e8197b395a/src/__tests__/book-screen.js#L186-L194

I just tried to reproduce the problem and can't 😬 It's possible that I've changed the test enough that I don't end up in the edge case situation. It's also possible that my deps have changed to side-step the edge case as well. I'm just not certain.

Basically, the problem stems from scheduling some timers in the scheduler when fake timers are enabled, then switching back to real timers without advancing all of the timers for the scheduler's callbacks. This happens in the first test. Then when the next test runs it attempts to schedule callbacks which are never run. I *think* they're not run because the scheduler is still waiting for the first callbacks to run which never happens because they were scheduled with the fake timer.

Comment (eps1lon) [reply src/flush-microtasks.js:92]: > Basically, the problem stems from scheduling some timers in the scheduler when fake timers are enabled, then switching back to real timers without advancing all of the timers for the scheduler's callbacks.

I was about to ask why they don't automatically advance until all are finished but I guess it isn't that obvious what we mean by "finished" in the case of `setInterval`. Definitely need to look into it and start running at least all the timeouts before restoring fake timers if that isn't already the case.

Comment (kentcdodds) [reply src/flush-microtasks.js:92]: Yeah, part of me feels like this problem should be solved in either the scheduler or jest 🤔

Comment (eps1lon) [reply src/flush-microtasks.js:92]: "infinite timers" is also something I haven't considered: https://jestjs.io/docs/en/timer-mocks.

If we're recommending to cleanup between every test we should probably start recommending to clean up timers as well (`runOnlyPendingTimers`/`clearAllTimers` and `useRealTimers`) (and before testing-library cleanup):

- beforeEach: useFakeTimers
- test: render()
- afterEach: clearAllTimers
- afterEach: useRealTimers
- afterEach: cleanup

The problem right now is that you can't achieve that order with automatic cleanup.

This ensures that only the code you can "see" in your test uses fake timers and not unrelated cleanup code from third-party libraries that isn't aware of what kind of timers you're using.

Comment (kentcdodds) [reply src/flush-microtasks.js:92]: it seems to me like the best course of action would be to add a page in the documentation explaining the best way to use fake timers with testing library.

Comment (MatanBobi) [reply src/flush-microtasks.js:92]: > Basically, the problem stems from scheduling some timers in the scheduler when fake timers are enabled, then switching back to real timers without advancing all of the timers for the scheduler's callbacks. This happens in the first test. Then when the next test runs it attempts to schedule callbacks which are never run. I _think_ they're not run because the scheduler is still waiting for the first callbacks to run which never happens because they were scheduled with the fake timer.

@kentcdodds 
This was exactly the case in https://github.com/testing-library/react-testing-library/issues/738,
the scheduler used by react-dom in the test was using fake timers and since they weren't advanced, the next `scheduleCallback` we've added in the cleanup step was stuck and couldn't schedule the callback since a callback wasn't flushed before.

I agree that adding a page in the documentation to explain the usage of fake timers in testing library will be good. 
I don't mind working on this one if we think that's the best approach :)

Comment (kentcdodds) [reply src/flush-microtasks.js:92]: I think that's the best we can do, so let's do it 👍

Comment (MatanBobi) [reply src/flush-microtasks.js:92]: @kentcdodds, @eps1lon, I'm working on documenting this but having some second thoughts:
using `clearAllTimers` in `afterEach` may be problematic in some cases, for example if react's scheduler uses fake timers within our test and they aren't advanced, the callback won't be called and the booleans they count on won't be updated.. This might not be a problem though since it's at the cleanup stage.. 
Wdyt?

Comment (kentcdodds) [reply src/flush-microtasks.js:92]: Could we recommended `advanceAllTimers`?

Comment (eps1lon) [reply src/flush-microtasks.js:92]: > advanceAllTimers

Did you mean `runAllTimers`? We probably don't want to do this because [`runAllTimers` can cause infinite loops](https://jestjs.io/docs/en/timer-mocks#run-pending-timers). `runOnlyPendingTimers` seems like it fits our use case since this would run all scheduled timers from `scheduler`.

Comment (kentcdodds) [reply src/flush-microtasks.js:92]: 😅 yeah, that's what I meant 👍👍

Comment (MatanBobi) [reply src/flush-microtasks.js:92]: Thanks @eps1lon, I wasn't aware of the infinite loop that `runAllTimers` can cause. Using `runOnlyPendingTimers` sounds like a good approach but I'm not sure all testing frameworks have that ability (I think jasmine only has `tick` functionality but I've never really used it). Anyways, my examples will use jest.

Comment (eps1lon) [reply src/flush-microtasks.js:92]: > Using runOnlyPendingTimers sounds like a good approach but I'm not sure all testing frameworks have that ability (I think jasmine only has tick functionality but I've never really used it). Anyways, my examples will use jest.

Which is why I don't think we should handle this in testing-library. Or at at least provide and entry point that makes no assumption about the testing framework.

Comment (MatanBobi) [reply src/flush-microtasks.js:92]: > Which is why I don't think we should handle this in testing-library. Or at at least provide and entry point that makes no assumption about the testing framework.

Your right.. Our plan wasn't to handle this within testing-library, just explain what should be done by our users within a new documentation page.. Do you think we shouldn't do that either?

Comment (kentcdodds) [reply src/flush-microtasks.js:92]: Let's definitely do that 👍
